src/colmap/estimators/bundle_adjustment.cc

```diff
@@ -329,17 +330,30 @@ ceres::Solver::Options BundleAdjuster::SetUpSolverOptions(
     const ceres::Problem& problem,
     const ceres::Solver::Options& input_solver_options) const {
   ceres::Solver::Options solver_options = input_solver_options;
+  if (VLOG_IS_ON(2)) {
+    solver_options.minimizer_progress_to_stdout = true;
+    solver_options.logging_type = ceres::LoggingType::PER_MINIMIZER_ITERATION;
+  }
+
   const bool has_sparse =
       solver_options.sparse_linear_algebra_library_type != ceres::NO_SPARSE;
 
-  // Empirical choice.
-  const size_t kMaxNumImagesDirectDenseSolver = 50;
-  const size_t kMaxNumImagesDirectSparseSolver = 1000;
-  const size_t num_images = config_.NumImages();
-  if (num_images <= kMaxNumImagesDirectDenseSolver) {
+  const int num_images = config_.NumImages();
+  if (num_images <= options_.max_num_images_direct_dense_solver) {
     solver_options.linear_solver_type = ceres::DENSE_SCHUR;
-  } else if (num_images <= kMaxNumImagesDirectSparseSolver && has_sparse) {
+  } else if (num_images <= options_.max_num_images_direct_sparse_solver &&
+             has_sparse) {
     solver_options.linear_solver_type = ceres::SPARSE_SCHUR;
+#if (CERES_VERSION_MAJOR >= 3 ||                                \
+     (CERES_VERSION_MAJOR == 2 && CERES_VERSION_MINOR >= 2)) && \
+    !defined(CERES_NO_CUDSS) && defined(CUDA_ENABLED)
+    if (options_.use_gpu) {
```

Comment:
I think that it might be worth experimentation/tuning to find problem size threshold to switch from CPU to GPU sparse backend, although it might be too sensitive to choice between suitesparse/eigen (until mkl sparse backend will be merged to ceres-solver).

Reply:
Yes, the current parameters are not tuned. It will be difficult to find universal thresholds here that generalize across different CPU/GPU models. It was a little easier before, as we switched between algorithms that all run on the CPU. As such, I decided to expose the parameters now explicitly through the option manager for the bundle adjuster and we might need to do the same for the bundle adjustment that runs as part of the mapper. For now, this feature is disabled by default, as I want to gain some more experience with it over the next weeks.